### PR TITLE
Another set of TODO 3.0 cleanups - this time mostly in crypto

### PR DESCRIPTION
**Pull request by t8m**
Please see the individual commits for details.

### CODE REVIEW

File: include/crypto/evp.h [hunk 1]
```diff
@@ -250,7 +249,7 @@ struct evp_md_st {
     int (*md_ctrl) (EVP_MD_CTX *ctx, int cmd, int p1, void *p2);
 
     /* New structure members */
-    /* TODO(3.0): Remove above comment when legacy has gone */
+    /* Above comment to be removed when #legacy has gone */
```
Comment (paulidale) [inline include/crypto/evp.h:252]: Why #legacy here but nowhere else?  I'd suggest removing the #

File: crypto/ec/ecx_meth.c [hunk 2]
```diff
@@ -334,7 +334,6 @@ static int ecd_ctrl(EVP_PKEY *pkey, int op, long arg1, void *arg2)
 static int ecx_set_priv_key(EVP_PKEY *pkey, const unsigned char *priv,
                             size_t len)
 {
-    /* TODO(3.0): We should pass a libctx here */
```
Comment (slontis) [inline crypto/ec/ecx_meth.c:337]: This one is potentially problematic and should probably not be part of this PR.

File: include/crypto/evp.h (hunk 1 above)
Comment (levitte) [reply include/crypto/evp.h:252]: There are other places as well. I started with that so it would be easy to grep when the time coms

Comment (levitte) [reply include/crypto/evp.h:252]: I agree, though, that it should be consistent. Contrary to @paulidale, I suggest adding the # in the other spots

Comment (paulidale) [reply include/crypto/evp.h:252]: It looks like 9 #legacy and 8 legacy.  I'm happy either way.

File: crypto/ec/ecx_meth.c (hunk 2 above)
Comment (t8m) [reply crypto/ec/ecx_meth.c:337]: I've added a code to get the libctx from the pkey's keymgmt if it is set. Not sure it will be so if this function is called though. However that is what we can do here.

File: include/crypto/evp.h (hunk 1 above)
Comment (t8m) [reply include/crypto/evp.h:252]: There are actually zillion places where legacy without # is used.

OK, when I look for legacy in a comment then I can still find about 340 occurrences. I'll change this to legacy without #. If anyone wants to do a global replacement with #legacy then this should be a different PR. But I do not think it is worth it.

File: crypto/evp/evp_enc.c [hunk 3]
```diff
@@ -597,7 +597,7 @@ int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
 
     return ret;
 
-    /* TODO(3.0): Remove legacy code below */
+    /* Code below to remove when legacy support is dropped. */
```
Comment (paulidale) [inline crypto/evp/evp_enc.c:600]: to **be removed**

File: crypto/evp/evp_enc.c [hunk 4]
```diff
@@ -657,7 +657,7 @@ int EVP_EncryptFinal_ex(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl)
...
```
Comment (paulidale) [inline crypto/evp/evp_enc.c:660]: to **be removed**

File: crypto/evp/evp_enc.c [hunk 5]
```diff
@@ -744,7 +744,7 @@ int EVP_DecryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
...
```
Comment (paulidale) [inline crypto/evp/evp_enc.c:747]: to **be removed**

File: crypto/evp/evp_enc.c [hunk 6]
```diff
@@ -879,7 +879,7 @@ int EVP_DecryptFinal_ex(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl)
...
```
Comment (paulidale) [inline crypto/evp/evp_enc.c:882]: to **be removed**

File: crypto/evp/evp_enc.c [hunk 7]
```diff
@@ -953,7 +953,7 @@ int EVP_CIPHER_CTX_set_key_length(EVP_CIPHER_CTX *c, int keylen)
         return ok > 0 ? 1 : 0;
     }
 
-    /* TODO(3.0) legacy code follows */
+    /* Code below to remove when legacy support is dropped. */
```
Comment (paulidale) [inline crypto/evp/evp_enc.c:956]: to **be removed**

File: crypto/evp/pmeth_lib.c [hunk 8]
```diff
@@ -963,9 +963,10 @@ static int evp_pkey_ctx_set1_octet_string(EVP_PKEY_CTX *ctx, int fallback,
         return -2;
     }
 
-    /* TODO(3.0): Remove this eventually when no more legacy */
+    /* Code below to remove when legacy support is dropped. */
```
Comment (paulidale) [inline crypto/evp/pmeth_lib.c:966]: to **be removed**

File: crypto/evp/pmeth_lib.c [hunk 9]
```diff
@@ -1051,11 +1052,11 @@ int EVP_PKEY_CTX_hkdf_mode(EVP_PKEY_CTX *ctx, int mode)
...
```
Comment (paulidale) [inline crypto/evp/pmeth_lib.c:1055]: to **be removed**

File: crypto/evp/pmeth_lib.c [hunk 10]
```diff
@@ -1099,9 +1100,10 @@ static int evp_pkey_ctx_set_uint64(EVP_PKEY_CTX *ctx, const char *param,
...
```
Comment (paulidale) [inline crypto/evp/pmeth_lib.c:1103]: to **be removed**

**Review by paulidale — APPROVED** [review]
One wording change in multiple places.  Approved with these addressed.

**Comment by t8m** [comment]
Fixup pushed addressing that.

**Review by paulidale — APPROVED** [review]
:)

File: crypto/ec/ecx_meth.c (hunk 2 above)
Comment (slontis) [reply crypto/ec/ecx_meth.c:337]: After tracking down where this is called from maybe this is not a good idea..

It seems to be called only from legacy paths for EVP_PKEY_new_raw_private_key_ex().
Provider keys would just use fromdata. The legacy path that uses this does
pkey = EVP_PKEY_new();
pkey_set_type(pkey, e,nidtype, strtype, -1, NULL) //where NULL is the keymanager.
pkey->ameth->set_priv_key().

So even though EVP_PKEY_new_raw_private_key_ex() passes a libctx, propq there is no passing of this down..
unless we had a variant of set_priv_key that passed the libctx,propq then it wont work..
@mattcaswell should we care about this?

Comment (slontis) [reply crypto/ec/ecx_meth.c:337]: looks like the set_priv_key was added in 1_1_1 so we cant just change it directly.

Comment (mattcaswell) [reply crypto/ec/ecx_meth.c:337]: I'm not sure why any of the code in this file needs a libctx. It's all legacy ameth/pmeth stuff which is typically not libctx aware. My guess is this was introduced at some intermediate stage where we didn't have all the encoder/decoder stuff etc in place yet???

Comment (t8m) [reply crypto/ec/ecx_meth.c:337]: I do not think the current state as being added in the PR is harmful, it is just most probably a dead code.

**Comment by t8m** [comment]
Merged to master. Thank you for the reviews.
